src/app/ReadClient.cpp

```diff
@@ -273,56 +330,73 @@ CHIP_ERROR ReadClient::SendReadRequest(ReadPrepareParams & aReadPrepareParams)
+        }
     }
 
     aAttributePathIBsBuilder.EndOfAttributePathIBs();
     return aAttributePathIBsBuilder.GetError();
 }
 
 CHIP_ERROR ReadClient::GenerateDataVersionFilterList(DataVersionFilterIBs::Builder & aDataVersionFilterIBsBuilder,
-                                                     DataVersionFilter * apDataVersionFilterList, size_t aDataVersionFilterListSize)
+                                                     const Span<AttributePathParams> & aAttributePaths,
+                                                     const Span<DataVersionFilter> & aDataVersionFilters, uint32_t & aNumber)
 {
-    for (size_t index = 0; index < aDataVersionFilterListSize; index++)
+    for (auto & filter : aDataVersionFilters)
     {
-        VerifyOrReturnError(apDataVersionFilterList[index].IsValidDataVersionFilter(), CHIP_ERROR_INVALID_ARGUMENT);
-        DataVersionFilterIB::Builder & filter = aDataVersionFilterIBsBuilder.CreateDataVersionFilter();
+        VerifyOrReturnError(filter.IsValidDataVersionFilter(), CHIP_ERROR_INVALID_ARGUMENT);
+
+        // if data version filter does not intersect with attribute path requests asks for, discard this filter
+        bool intersected = false;
+        for (auto & path : aAttributePaths)
+        {
+            if (path.IsAttributePathIntersect(filter))
+            {
+                intersected = true;
+                break;
+            }
+        }
+        if (!intersected)
+        {
+            continue;
+        }
+
+        DataVersionFilterIB::Builder & filterIB = aDataVersionFilterIBsBuilder.CreateDataVersionFilter();
         ReturnErrorOnFailure(aDataVersionFilterIBsBuilder.GetError());
-        ClusterPathIB::Builder & path = filter.CreatePath();
-        ReturnErrorOnFailure(filter.GetError());
-        ReturnErrorOnFailure(path.Endpoint(apDataVersionFilterList[index].mEndpointId)
-                                 .Cluster(apDataVersionFilterList[index].mClusterId)
-                                 .EndOfClusterPathIB()
-                                 .GetError());
-        VerifyOrReturnError(apDataVersionFilterList[index].mDataVersion.HasValue(), CHIP_ERROR_INVALID_ARGUMENT);
-        ReturnErrorOnFailure(
-            filter.DataVersion(apDataVersionFilterList[index].mDataVersion.Value()).EndOfDataVersionFilterIB().GetError());
+        ClusterPathIB::Builder & path = filterIB.CreatePath();
+        ReturnErrorOnFailure(filterIB.GetError());
+        ReturnErrorOnFailure(path.Endpoint(filter.mEndpointId).Cluster(filter.mClusterId).EndOfClusterPathIB().GetError());
+        VerifyOrReturnError(filter.mDataVersion.HasValue(), CHIP_ERROR_INVALID_ARGUMENT);
+        ReturnErrorOnFailure(filterIB.DataVersion(filter.mDataVersion.Value()).EndOfDataVersionFilterIB().GetError());
+        aNumber++;
     }
 
+    ReturnErrorOnFailure(aDataVersionFilterIBsBuilder.GetWriter()->UnreserveBuffer(kReservedSizeForTLVEncodingOverhead));
```

Comment:
If this call moved into the one callsite, then we would not need the boolean to keep track of whether we had done it.  We'd just always do it unconditionally after the if/else at the callsite...  Would be simpler that way.

Reply:
done, thanks